File: III.Ödev/src/Oyun.java
import java.util.StringTokenizer;

public class Oyun {
	Koloni[] koloniler;
    private int tur;
    

    public Oyun(int ac, String av) {
        tur = 0;
        koloniler = getKoloniler(ac, av);
    }

	public void deleteOyun(Oyun oyun, int sayi) {
        deleteKoloniler(sayi);
    }

    private Koloni[] getKoloniler(int arc, String av) {
        int ac = getAc(av);
        StringTokenizer tokenizer = new StringTokenizer(av, " ");
        Koloni[] koloniler = new Koloni[ac - 1];
        int i = 1;

        while (tokenizer.hasMoreTokens()) {
            int number = Integer.parseInt(tokenizer.nextToken());
            if (number > 0) {
            	int a = i+34;
                koloniler[i - 1] = new Koloni((char) (a), number);
                i++;
                
            }
        }

        return koloniler;
    }
    
    private int getAc(String str) {
        StringTokenizer tokenizer = new StringTokenizer(str, " ");
        int ac = 1;

        while (tokenizer.hasMoreTokens()) {
            int number = Integer.parseInt(tokenizer.nextToken());
            if (number > 0) {
                ac++;
            } else {
                System.out.println("Lütfen geçerli bir input giriniz! (sadece pozitif sayılar ve boşluk)");
                System.exit(1);
            }
        }

        return ac;
    }

    public void printKoloniler(Koloni[] koloniler) {
        System.out.println("Koloni\t\tPop.\t\tYemek\t\tKazanma\t\tKaybetme");
        for (int i = 0; i < koloniler.length; i++) {
            koloniler[i].printKoloni();
        }
    }

    boolean isGood(Koloni[] koloniler) {
        int kontrol = 0;

        for (int i = 0; i < koloniler.length; i++) {
            if (koloniler[i].getPopulasyon() == 0) {
                kontrol++;
            }
        }

        return koloniler.length - 1 > kontrol;
    }

    public void uretKoloniler(Koloni[] koloniler) {
        for (int i = 0; i < koloniler.length; i++) {
            if (koloniler[i].getPopulasyon() > 0) {
            	if (koloniler[i].getUretimTip() == 'a') {
                    koloniler[i].setYemek(koloniler[i].getUretim().uret(koloniler[i].getPopulasyon()));
                }

                if (koloniler[i].getUretimTip() == 'b') {
                    koloniler[i].setYemek(koloniler[i].getUretim().uret(koloniler[i].getPopulasyon()));
                }
            }
        }
    }

    public void turetKoloniler(Koloni[] koloniler) {
        for (int i = 0; i < koloniler.length; i++) {
            if (koloniler[i].getPopulasyon() > 0) {
                koloniler[i].setPopulasyon(koloniler[i].getPopulasyon() + koloniler[i].getPopulasyon() / 5);
            }
        }
    }

    public void yedirKoloniler(Koloni[] koloniler) {
        for (int i = 0; i < koloniler.length; i++) {
            if (koloniler[i].getPopulasyon() > 0) {
                koloniler[i].setYemek(koloniler[i].getYemek() - koloniler[i].getPopulasyon() * 2);
                if (koloniler[i].getYemek() < 0) {
                    koloniler[i].setPopulasyon(0);
                    koloniler[i].setYemek(0);
                }

            }
        }
    }

    public void savastirKoloniler(Koloni[] koloniler) {
        for (int i = 0; i < koloniler.length; i++) {
            for (int j = i + 1; j < koloniler.length; j++) {
                if (koloniler[i].getPopulasyon() > 0 && koloniler[j].getPopulasyon() > 0) {
                    koloniler[i].savas(koloniler[j]);
                }
            }
        }
    }

    private void deleteKoloniler(int sayi) {
        for (int i = 0; i < sayi; i++) {
            koloniler[i].yoket();
        }
    }
}

File: II.Ödev/src/oyun.c
#include "oyun.h"

void    deleteOyun(Oyun oyun, int sayi)
{
    oyun->deleteKoloniler(oyun->koloniler, sayi);
    free(oyun);
}

Oyun  newOyun(int *ac, char *av)
{
    Oyun this = (Oyun)malloc(sizeof(struct OYUN));

    this->tur = 0;
    this->getKoloniler = &getKoloniler;
    this->printKoloniler = &printKoloniler;
    this->uretKoloniler = &uretKoloniler;
    this->turetKoloniler = &turetKoloniler;
    this->yedirKoloniler = &yedirKoloniler;
    this->savastirKoloniler = &savastirKoloniler;
    this->deleteOyun = &deleteOyun;
    this->deleteKoloniler = &deleteKoloniler;

    this->koloniler = this->getKoloniler(ac, av);

    return this;
}

int getAc(const char *str)
{
    const char *delim = " ";
    char *token;
    char *str_copy = strdup(str);
    int ac = 1;

    if (str_copy == NULL) {
        printf("malloc hata\n");
        exit(1);
    }

    token = strtok(str_copy, delim);

    while (token != NULL)
    {
        int number = atoi(token);
        if (number > 0)
        {
            ac++;
        }

        else
        {
            printf("lutfen gecerli bir input giriniz!(sadece pozitif sayilar ve bosluk)\n");
            exit(1);

        }

        token = strtok(NULL, delim);
    }

    free(str_copy);
    return (ac);
}

Koloni *getKoloniler(int *arc, char *av)
{
    Koloni *koloniler;

    *arc = getAc(av);
    int ac = *arc;
    const char *delim = " ";
    char *token;
    char *str_copy = strdup(av);
    
    koloniler = (Koloni *)malloc((sizeof(Koloni) + 1) * (ac - 1));
    token = strtok(str_copy, delim);

    int i = 1;
    while (token != NULL)
    {
        int number = atoi(token);
        if (number > 0)
        {
            koloniler[i - 1] = newKoloni(i + 33, number);
            i++;
        }

        token = strtok(NULL, delim);
    }
    koloniler[ac - 1] = NULL;
    
    return koloniler;
}

void printKoloniler(Koloni *koloniler, int sayi)
{
    printf("Koloni\t\tPop.\t\tYemek\t\tKazanma\t\tKaybetme\n");
    for (int i = 0; i < sayi; i++)
    {
        koloniler[i]->printKoloni(koloniler[i]);
    }
}

boolean isGood(Koloni *koloniler, int sayi)
{
    int kontrol = 0;

    for (int i = 0; i < sayi; i++)
    {
        if (koloniler[i]->populasyon == 0)
        {
            kontrol++;
        }
    }

    if (sayi - 1 <= kontrol)
    {
        return false;
    }
    return true;
}

void    uretKoloniler(Koloni *koloniler, int sayi)
{
    for (int i = 0; i < sayi; i++)
    {
        if (koloniler[i]->populasyon > 0)
        {
            if (koloniler[i]->uretimTip == 'a')
            {
                koloniler[i]->yemek += ((UretimA)(koloniler[i]->uretim))->uret(koloniler[i]->populasyon);
            }

            if (koloniler[i]->uretimTip == 'b')
            {
                koloniler[i]->yemek += ((UretimB)(koloniler[i]->uretim))->uret(koloniler[i]->populasyon);
            }

        }

        
    }
    
}

void    turetKoloniler(Koloni *koloniler, int sayi)
{
    for (int i = 0; i < sayi; i++)
    {
        if (koloniler[i]->populasyon > 0)
        {
            koloniler[i]->populasyon += koloniler[i]->populasyon / 5;

        }
    }
    
}

void    yedirKoloniler(Koloni *koloniler, int sayi)
{
    for (int i = 0; i < sayi; i++)
    {
        if (koloniler[i]->populasyon > 0)
        {
            koloniler[i]->yemek -= koloniler[i]->populasyon * 2;
            if (koloniler[i]->yemek < 0)
            {
                koloniler[i]->populasyon = 0;
                koloniler[i]->yemek = 0;
            }

        }
    }
    
}



void    savastirKoloniler(Koloni *koloniler, int sayi)
{
    for (int i = 0; i < sayi; i++)
    {
        for (int j = i + 1; j < sayi; j++)
        {
            if (koloniler[i]->populasyon > 0 && koloniler[j]->populasyon > 0)
            {
                
                koloniler[i]->savas(koloniler[i], koloniler[j]);
            }
            
        }
        
    }
    
}

void    deleteKoloniler(Koloni *koloniler, int sayi)
{
    for (int i = 0; i < sayi; i++)
    {
        koloniler[i]->yoket(koloniler[i]);
    }
    free(koloniler);
    
}